### DISCUSSION TITLE
atmega328pb support

### DISCUSSION THREAD

**Discussion by alxgarza** [dbuezas/arduino-web-timers#33] [Ideas]
Any chance of adding:
- custom clock frequency
- Timer 4 (atmega328pb)

**Comment by dbuezas** [comment]
I don't own any 328pb boards, so I could add it if you promise to test it :)

Regarding a custom clock frequency, this was asked for already, but it is a bit more work and I don't understand the value of this (particularly b/c it would have no effect in the generated code). If you have a convincing use case I can give it a go soon

**Comment by alxgarza** [comment]
Glad to test timer 4.

The only value for setting a custom Clock Frequency is to calculate the final timer frequency.

Also, setting the target interrupt frequency would be a nice addition.

**Reply by dbuezas** [reply]
I'll have to take a look at the datasheet then. I assume it is similar to timer1

**Reply by dbuezas** [reply]
oh, the 328pb also has a timer3!

**Reply by alxgarza** [reply]
This is a nice document that shows the differences between ATmega328/P and ATmega328PB
https://www.pololu.com/file/0J1464/Atmel-42559-Differences-between-ATmega328P-and-ATmega328PB_ApplicationNote_AT15007.pdf

**Reply by dbuezas** [reply]
The doc doesn't have any specifics on the new timers unfortunately.
It does tho talk about a new peripheral: Output Compare Modulator. Which sounds intriguing. I won't add it yet.

In the datasheet there is what looks like an errata. In mode 8, it looks like the timer 3 uses icr1 as top instead of icr3. I'll assume this is a mistake, but let me know if you find it doesn't work as expected:
![image](https://github.com/dbuezas/arduino-web-timers/assets/777196/a2519af6-541e-4085-99e9-4004dbe059dc)
I didn't find any more differences between timers 3 and 4 and the timer 1 in the 328p.

**Comment by dbuezas** [comment]
Allright, 328pb added. Let me know if you find any oddities, as it is a direct copy of the 328p but with 2 extra timers adapted from timer1

**Reply by dbuezas** [reply]
It's deploying and it looks like it may take a couple of minutes for the github action to get a runner.
You can check the state here: https://github.com/dbuezas/arduino-web-timers/actions/runs/5316219388/jobs/9625502793
Once it is deployed, you'll see there is a new chip in the list: atemaga328pb

**Reply by alxgarza** [reply]
Excellent!

**Comment by dbuezas** [comment]
@alxgarza, do you have any feedback? I don't have 328pb chips to test this on

**Reply by alxgarza** [reply]
Not yet. I'm out of the country.

**Reply by alxgarza** [reply]
As far I can tell, it works properly.
